compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -906,8 +945,22 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(implicit ctx: Context) {
         val selType = if (sel.isEmpty) wideSelType else sel.tpe
         reduceInlineMatch(sel, selType, cases.asInstanceOf[List[CaseDef]], this) match {
           case Some((caseBindings, rhs0)) =>
-            val (usedBindings, rhs1) = dropUnusedDefs(caseBindings, rhs0)
-            val rhs = seq(usedBindings, rhs1)
+            // drop type ascriptions/casts hiding internal types (which are now aliases after reducing the match)
+            val rhs1 = rhs0 match {
+              case Block(stats, t) if t.pos.isSynthetic =>
```

Comment:
Are we sure this covers only blocks generated by the reduceMatch? Or could this be a synthetic block that was previously computed in which case the operations would be unsound? Can we find a more robust criterion for this?

Reply:
As I understand it, `reduceMatch` exposes the final block that could be inlined. I think the code is sound as long as long as type ascriptions/casts at the absolute end of blocks are inserted only to hide internal types and support GADT bounds. Note that even if some types still need to be hidden, casts will be re-inserted by `typedExpr` below.

If that's necessary, we could add annotations to casts intended to hide internal types, and look for them here. I believe we cannot simply avoid inserting these casts, since then the check for escaping internal aliases would complain when visiting inline method definitions.